spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java

```diff
@@ -303,16 +316,25 @@ protected boolean shouldCopyRequestHeaders() {
 
 		private final MessageHandler delegate;
 
+		private final boolean useNativeEncoding;
+
 		private SendingHandler(MessageHandler delegate, boolean embedHeaders,
-				String[] headersToEmbed) {
+				String[] headersToEmbed, boolean useNativeEncoding) {
 			this.delegate = delegate;
 			this.setBeanFactory(AbstractMessageChannelBinder.this.getBeanFactory());
 			this.embedHeaders = embedHeaders;
 			this.embeddedHeaders = headersToEmbed;
+			this.useNativeEncoding = useNativeEncoding;
 		}
 
 		@Override
 		protected void handleMessageInternal(Message<?> message) throws Exception {
+			Message<?> messageToSend = (this.useNativeEncoding) ? getMessageBuilderFactory().withPayload(message.getPayload()).build() :
```

Comment:
We shouldn't strip the headers if native encoding is being used. They may contain information useful for the sender (e.g. the partition id in partitioned scenarios, the key for Kafka, etc). if native encoding is being used, we should just use the incoming message as is.